slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -224,6 +225,113 @@ void SLAMServiceImpl::SetUpMapBuilder() {
+    auto painted_surface = painted_slices.surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+    int stride = cairo_image_surface_get_stride(painted_surface);
+    auto format = cairo_image_surface_get_format(painted_surface);
+
+    // cartographer format is ARGB32(but RGB is still first?)
+    bool format_check = format == cartographer::io::kCairoFormat;
+
+
+    std::cout << "width_px: " << width << std::endl;
+    std::cout << "height_px: " << height << std::endl;
+    std::cout << "stride_bytes: " << stride << std::endl;
+    std::cout << "format_check: " << format_check << std::endl;
+    std::cout << "format: " << format << std::endl;
+
+    //total number of bytes, 32 bits per pixel
+    size_t size_data = width*height*4;
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+    
+    std::stringbuf data_buffer;
+
+    int num_points = 0;
+    // not working yet, need to reduce resolution to hit 32 MB on our PCD
+    int scale = size_data*4/100000.f/32.f*4;
+
+    //i=i+4+12, +4 for bytes per pixel, +12 to write every 4th pixel
+    for (int i=0;i<size_data;i=i+4+12) {
+        int rgb = 0;
+        rgb = rgb | ((int)data_vect[i+0] << 16);
+        rgb = rgb | ((int)data_vect[i+1] << 8);
+        rgb = rgb | ((int)data_vect[i+2] << 0);
+
+        //skip pixels that are not in our map(black/past walls)
+        if(rgb == 6710886)
+        continue;
+
+        num_points++;
+
+        //mod math cuz its the best
+        int pixel_index = i/4;
+        int pixel_x = pixel_index/width;
+        int pixel_y = pixel_index%width;
+
+        //kPixelSize is .01, unsure of units but assumed converts to meters
+        // based on PaintSubmapSlices() and DrawPoseOnSurface()
+        float x_pos = pixel_x*kPixelSize;
+        float y_pos = pixel_y*kPixelSize;
+
+        //Magic number that gets a map to show up
+        float x_pos2 = x_pos/1000.f;
+        float y_pos2 = y_pos/1000.f;
+
+        //debuging
+        if((num_points > 1582734)&&(num_points < 1582739)){
+        std::cout << "pixel_index: " << pixel_index << std::endl;
+        std::cout << "pixel_x: " << pixel_x << std::endl;
+        std::cout << "pixel_y: " << pixel_y << std::endl;
+        std::cout << "x_pos: " << x_pos << std::endl;
+        std::cout << "y_pos: " << y_pos << std::endl;   
+        std::cout << "Working x_pos: " << x_pos2 << std::endl;
+        std::cout << "Working y_pos: " << y_pos2 << std::endl;
+        }
+        
+        //write points to buffer, same usage as before
+        float z_pos = 0;
+        data_buffer.sputn((const char *)&x_pos, 4);
+        data_buffer.sputn((const char *)&y_pos, 4);
+        data_buffer.sputn((const char *)&z_pos, 4);
+        data_buffer.sputn((const char *)&rgb, 4);
```

Comment:
will have double check to see what Jeremy did, as the PCD format expects an RGB value from what I understand and transforms that into a color

Reply:
Kk so yeah need to translate the RGB values pulled from cartographer into a value from 0 to 100, so i'll start looking into that

Reply:
The color and value are separate fields in PointCloud Data: https://github.com/viamrobotics/rdk/blob/3db1629d010a55560b5dc378697243eb216c782a/pointcloud/point.go#L38. The probability should be stored in the value, not in the color.

Reply:
but we read in the color data as an int32 and extract that back into a RGBA color point. the PCD does not have any separate channel for a single integer/int8 value.

Reply:
Got it. I assumed we did, based on the fact that Value is part of Data. It looks like the pcd file format is flexible for including more fields: https://pointclouds.org/documentation/tutorials/pcd_file_format.html. But it may not be worth adding flexibility in how we parse pcds just for the probability.

I suggest just making sure you and @jeremyhyde-viam are on the same page about how the probability will be included in the pcd and consumed by rdk.